tools/SourceKit/lib/SwiftLang/SwiftASTManager.cpp

```diff
@@ -347,13 +347,14 @@ struct CacheKeyHashInfo<ASTKey> {
 } // namespace swift
 
 struct SwiftASTManager::Implementation {
-  explicit Implementation(SwiftLangSupport &LangSupport)
-      : EditorDocs(LangSupport.getEditorDocuments()),
-        Stats(LangSupport.getStatistics()),
-        RuntimeResourcePath(LangSupport.getRuntimeResourcePath()) {}
-
-  SwiftEditorDocumentFileMap &EditorDocs;
-  SwiftStatistics &Stats;
+  explicit Implementation(
+      std::shared_ptr<SwiftEditorDocumentFileMap> EditorDocs,
+      std::shared_ptr<SwiftStatistics> Stats, StringRef RuntimeResourcePath)
+      : EditorDocs(EditorDocs), Stats(Stats),
+        RuntimeResourcePath(RuntimeResourcePath) {}
+
+  std::shared_ptr<SwiftEditorDocumentFileMap> EditorDocs;
+  std::shared_ptr<SwiftStatistics> Stats;
```

Comment:
Would it simplify things to make `SwiftLangSupport` a shared_ptr that `SwiftASTManager` weakly references ? Then `SwiftASTManager` wouldn't have need to have separate references for the things that `SwiftLangSupport` owns.

Reply:
That was my original approach, but I changed my mind.  It works, but it makes the code more complex at every use site (having to lock the weak_ptr then having to handle nullptr where today it's never null).  By storing references only to what we need there are a lot fewer weak pointers to worry about.